### ISSUE TITLE
Bug when rendering tables

### ISSUE THREAD

**Issue by fabiospampinato** [showdownjs/showdown#666]
The following table isn't recognized as a table:

```markdown
|   | ico | emoji | ico | emoji |
| - | --- | ----- | --- | ----- |
| [top](#table-of-contents) | :bowtie: | `:bowtie:`  |  :smile: | `:smile:` |
| [top](#table-of-contents) | :laughing: | `:laughing:`  |  :blush: | `:blush:` |
```

But this is:

```markdown
|   | ico | emoji | ico | emoji |
| -- | --- | ----- | --- | ----- |
| [top](#table-of-contents) | :bowtie: | `:bowtie:`  |  :smile: | `:smile:` |
| [top](#table-of-contents) | :laughing: | `:laughing:`  |  :blush: | `:blush:` |
```

Basically it looks like showdown is expecting multiple hyphens.

**Comment by tivie** [comment]
Yes, to avoid ambiguous situations, tables syntax requires at least 2 hyphens as separators.

**Comment by fabiospampinato** [comment]
@tivie Can you provide an example of one of those ambiguous situations? 🤔 

Anyway given that GitHub itself doesn't seem to behave like that perhaps it would be better to align showdown with it.

**Comment by tivie** [comment]
> @tivie Can you provide an example of one of those ambiguous situations?

Sure.

```
 a | b
 - | -
 c | d
```

should this be rendered as a table? or as a paragraph followed by a list?

**Comment by fabiospampinato** [comment]
Right, but I have a few problems with this:

1. The table I posted doesn't have such ambiguity, since the divider row starts with a pipe and not with an hyphen, this enforced double-hyphen rule seems to break some perfectly valid and unambiguous tables to me, right?

2. This syntax for writing ables isn't specced in the original Markdown spec, and since showdown supposedly supports the `gfm` spec, which defines a syntax for tables, I guess tables should be rendered the same way they are rendered here on GitHub, otherwise those are not `gfm` tables but showdown-formatted-markdown tables, whatever that means, no?

**Comment by tivie** [comment]
TL;DR;

It's a pain because I think Showdown's implementation is better than GFM's and GFM Implementation will eventually be changed (maybe?). We need a lot of changes to make tables fully GFM compliant and some are clearly backwards incompatible (such as blank line after table requirement).
This particular fix you point out requires a lot of code change.

---


> The table I posted doesn't have such ambiguity, since the divider row starts with a pipe and not with an hyphen this enforced double-hyphen rule seems to break some perfectly valid and unambiguous tables to me, right?
Yes, there is no ambiguity in your specific example, since it clearly seems a table 😅. But since our spec states that leading and trailing pipes are optional, it becomes ambiguous if you remove the optional leading and trailing pipes.
Also, there are more examples of ambiguity in GFM.

```
| |
|-|
4x0r
```

GFM parses this like this:

| |
|-|
4x0r


Of course this is super edge case and a super weird way of writing "pipe haxor", but you get the point.
This also hints other cases which, IMHO, are ambiguous in GFM Table spec, although they are not directly related to the number of dashes...

``` 
| foo |
|-|
bar
baz
bazinga
```

| foo |
|-|
bar
baz
bazinga

> This syntax for writing tables isn't specced in the original Markdown spec
Yes, tables are not a standard syntax of markdown. Showdown's implementation is based off the commonmark tables syntax, which is not part of core spec yet (it's just a proposed extension), but it's reasonable to expect it will become part of core sooner or later.

> since showdown supposedly supports the gfm spec, which defines a syntax for tables, I guess tables should be rendered the same way they are rendered here on GitHub otherwise those are not gfm tables but showdown-formatted-markdown tables, whatever that means, no?
That is technically true. However there are some issues that make me kind of resistant to the change...
 - GFM became commonmark compliant a while ago. That means it should support the full commonmark testcase (whihc it does). But since tables syntax is not core spec yet, GFM retained their old RedCarpet implementation. 
 - If we switch to GFM tables, we would need to change our spec and later change it again, when commonmark spec eventually moves tables to core spec.
 - GFM implementation is slower and more complex, since it has a couple of extra rules to prevent ambiguity.
 - Showdown's set of rules are faster and simpler, since you don't need to test for "lists".
 - It might be confusing for users, since they might expect that leading and trailing.
 - Lastly, there are other changes that we would need to make in order to be fully compliant with GFM tables, namely:
    - GFM doesn't need any characters inside `|`, showdown needs at least one character.
    - Showdown requires an empty line after tables, GFM doesn't.
    - In GFM, `|` are mandatory or optional on a case by case basis (basically GFM turns everything into tables except if it's turned into something else, like a list). In showdown are always optional EXCEPT in one column tables, in which they are ALWAYS required in ALL lines.

**Comment by fabiospampinato** [comment]
Thanks for the detailed explanation, let's hope GitHub's implementation becomes less ambiguous and aligns itself more to commonmark's table syntax then 👍

**Comment by connorjs** [comment]
I came across this issue today (thanks for the detailed commentary!) when hitting the "single hyphen" case myself. Do you know how we "ask GitHub" to clarify their table spec? I could not find any associated repo from some personal searching.

[GFM spec section](https://github.github.com/gfm/#delimiter-row) under discussion

> The delimiter row consists of cells whose only content are hyphens (-), and optionally, a leading or trailing colon (:), or both, to indicate left, right, or center alignment respectively.

Which has an example later in the [next section](https://github.github.com/gfm/#example-199) showing one hyphen.

> ```
> | abc | defghi |
> :-: | -----------:
> bar | baz
> ```

However, this "single hyphen" does seem like a "cheeky" example when viewed through the lens of this discussion thread, because it does not have ambiguity due to the colons.

**Goal**: I'm looking how to reconcile that `remark` (via `remark-gfm`) and `remarkable` allow single hyphen tables, but `showdown` does not (even though source documents both say GFM).

**Comment by bensquire** [comment]
@connorjs 

> I'm looking how to reconcile that remark (via remark-gfm) and remarkable allow single hyphen tables, but showdown does not (even though source documents both say GFM).

Just hit this very exact scenario myself, did you discover anything?
